akka-stream/src/main/scala/akka/stream/scaladsl/StreamConverters.scala

```diff
@@ -90,4 +96,99 @@ object StreamConverters {
+      .map(state ⇒ state().finish())
+      .toMat(Sink.head)(Keep.right).named("javaCollectorSink")
+
+  /**
+    * Creates a sink which materializes into ``Future`` which will be completed with result of the Java 8 Collector transformation
+    * and reduction operations. This allows to use Java 8 streams transformations for reactive streams. Collector is triggering
+    * demand downstream. Elements emitted through the stream will be accumulated into a mutable result container, optionally transformed
+    * into a final representation after all input elements have been processed. Collector can also do reduction
+    * at the end. Reduction processing performs in parallel based on graph ``Balance``.
+    *
+    * Note that a flow can be materialized multiple times, so the function producing the ``Collector`` must be able
+    * to handle multiple invocations.
+    */
+  def javaCollectorParallelUnordered[T, R](parallelism: Int)(collectorFactory: () ⇒ java.util.stream.Collector[T, _ <: Any, R]): Sink[T, Future[R]] = {
+    if (parallelism == 1) javaCollector[T, R](collectorFactory)
+    else {
+      Sink.fromGraph(GraphDSL.create(Sink.head[R]) { implicit b ⇒
+        sink ⇒
+          import GraphDSL.Implicits._
+          val collector = collectorFactory().asInstanceOf[Collector[T, Any, R]]
+          val balance = b.add(Balance[T](parallelism))
+          val merge = b.add(Merge[() ⇒ CollectorState[T, R]](parallelism))
+
+          for (i ← 0 until parallelism) {
+            val worker = Flow[T]
+              .fold(() => new CollectorState(collector)) { (state, elem) ⇒ () ⇒ state().update(elem) }
+              .async
+
+            balance.out(i) ~> worker ~> merge.in(i)
+          }
+
+          merge.out
+            .fold(() => new ReducerState(collector)) { (state, elem) ⇒ () ⇒ state().update(elem().accumulated) }
+            .map(state => state().finish()) ~> sink.in
+
+          SinkShape(balance.in)
+      }).named("javaCollectorAsyncUnorderedSink")
+    }
+  }
+
+  /**
+    * Creates a sink which materializes into Java 8 ``Stream`` that can be run to trigger demand through the sink.
+    * Elements emitted through the stream will be available for reading through the Java ``Stream``.
+    *
+    * The Java 8 ``Stream`` will be ended when the stream flowing into this ``Sink`` completes, and closing the Java
+    * ``Stream`` will cancel the inflow of this ``Sink``.
+    *
+    * Java Stream throws exception in case reactive stream failed.
+    */
+  def asJavaStream[T](): Sink[T, java.util.stream.Stream[T]] = {
+    Sink.fromGraph(new QueueSink[T]()).mapMaterializedValue(queue ⇒ StreamSupport.stream(
+      Spliterators.spliteratorUnknownSize(new java.util.Iterator[T] {
+        var nextElementFuture: Future[Option[T]] = queue.pull()
+        var nextElement: Option[T] = null
+
+        override def hasNext: Boolean = {
+          nextElement = Await.result(nextElementFuture, Inf)
```

Comment:
@patriknw are you thinking about having it non blocking or just another dispatcher?